doc/developer/platform/formalism.md

```diff
@@ -0,0 +1,196 @@
+
+## Storage
+
+The Storage layer presents a narrow API about which it makes guarantees.
+It likely has a more verbose diagnostic API that describes its state, which should be a view over the results of these commands.
+
+*   `Create(id, description) -> (ReadCapability, WriteCapability)`: binds to `id` the TVC described by `description`.
+
+    The command returns capabilities naming `id`, with frontiers set to initial values chosen by the Storage layer.
+
+    The collection associated with `id` is based on `description`, but it is up to the Storage layer to determine what the source's contents will be.
+    A standarad example is a Kafka topic, whose contents are "added" to the underlying collection as they are observed.
+    The Storage layer only needs to ensure that it can reliably produce the updates between `since` and `upper`, and otherwise the nature of the content is up for discussion.
+    It is an error to re-use a previously used `id`.
+
+*   `SubscribeAt(ReadCapability(id, frontier))`: returns a snapshot of `id` at `frontier`, followed by an ongoing stream of subsequent updates.
+
+    This command returns the contents of `id` as of `frontier` once they are known, and updates thereafter once they are known.
+    The snapshot and stream contain in-line statements of the subscription's `upper` frontier: those times for which updates may still arrive.
+    The subscription will produce exactly correct results: the snapshot is the TVCs contents at `frontier`, and all subsequent updates occur at exactly their indicated time.
+    This call may block, or not be promptly responded to, if `frontier` is greater or equal to the current `upper` of `id`.
+    The subscription can be canceled by either endpoint, and the recipient should only downgrade their read capability when they are certain they have all data through the frontier they would downgrade to.
+
+    A subscription can be constructed with additional arguments that change how the data is returned to the user.
+    For example, the user may ask to not receive the initial snapshot, rather that receive and then discard it.
+    For example, the user may provide filtering and projection that can be applied before the data are transmitted.
+
+*   `UpdateAndDowngrade(WriteCapability(id, frontier), updates, new_frontier)`: applies `updates` to `id` and downgrades `frontier` to `new_frontier`.
+
+    All times in `updates` must be greater or equal to `frontier` and not greater or equal to `new_frontier`.
+
+    It is probably an error to call this with `new_frontier` equal to `frontier`, as it would mean `updates` must be empty.
+    This is the analog of the `INSERT` statement, though `updates` are signed and general enough to support `DELETE` and `UPDATE` at the same time.
+
+*   `AcquireCapabilities(id) -> (ReadCapability, WriteCapability)`: provides capabilities for `id` at its current `since` and `upper` frontiers.
```

Comment:
To check my understanding, is this basically for ADAPTOR restarts and a new (re)started ingestd or dataflowd would be getting its capabilities through the clone and share method?

Reply:
I could see this even being used as the main way it gets capabilities, as its pattern is often "generally, I don't care to hold back compaction, but when I need to perform a transaction I will want to acquire and then temporarily hold read capabilities to make sure I provide consistent answers". I think there are few reasons the coordinator wants to hold on to capabilities at the moment; other than transactions, it is mostly using them to 1. implement compaction windows, and 2. propagate constraints from EOSinks back to sources.

Internal mechanism like dataflow creation would result in capability clones, with the capability most likely held in the durable state of the COMPUTE controller, and potentially communicated to `dataflowd` instances. It's not clear at the moment that the `dataflowd` instances need to have the capability themselves (perhaps the controller initiates the subscription using a capability it holds, and then passes the subscription identifier along to the `dataflowd` instance.

Writing code at the moment, it feels more likely than not that most of these capabilities stay within the bounds of the controller, and are used to communicate constraints that are decentralized more than they are distributed, if that makes sense.